lib/go/thrift/header_context.go

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package thrift
+
+import (
+	"context"
+)
+
+// See https://godoc.org/context#WithValue on why do we need the unexported typedefs.
+type (
+	headerKey     string
+	headerKeyList int
+)
+
+// Values for headerKeyList.
+const (
+	headerKeyListRead headerKeyList = iota
+)
+
+// SetContextHeader sets a header in the context.
+func SetContextHeader(ctx context.Context, key, value string) context.Context {
```

Comment:
`SetHeader` sounds better, no need to mention `Context` in the function name. Same goes for all the other functions in this file.

Reply:
@dcelasun My main concern is that we have functions sound similar, like [`thrift.(*THeaderProtocol).SetWriteHeader`](https://godoc.org/github.com/apache/thrift/lib/go/thrift#THeaderProtocol.SetWriteHeader). Although there's no conflicts this might look confusing and making it more explicit might be better. Thoughts?

Reply:
I did it in a fixup commit. If we decided that we do want the rename I'll squash the branch.

Reply:
Since the other functions belong to `THeaderProtocol` and these new ones are on the package level, I don't think there'll be much confusion.

If you have any other changes you'd like to do, please go ahead. Otherwise I'm gonna squash & merge.

Reply:
I squashed them. Thanks for the review!